ui/components/src/Legend/ListLegend.tsx

```diff
@@ -11,40 +11,155 @@
+  // https://github.com/bvaughn/react-window#can-i-add-padding-to-the-top-and-bottom-of-a-list
+  const LIST_PADDING = parseInt(theme.spacing(1), 10);
+
   // show full labels on hover when there are many total series
   const truncateLabels = items.length > 5;
-  return (
-    <List>
-      {items.map((item) => (
+
+  // Gets the row height for a given item to enable the virtualized list to
+  // render the row properly.
+  function getRowHeight(index: number) {
+    const currentHeight = rowHeights.current[index];
+    return currentHeight ?? DEFAULT_ROW_HEIGHT;
+  }
+
+  // Set the height for a given item to enable the virtualized list to
+  // adjust to size changes.
+  function setRowHeight(index: number, size: number) {
+    // Tell the virtualized list that items changed size and need to be
+    // re-evaluated.
+    listRef.current?.resetAfterIndex(0);
+    rowHeights.current = { ...rowHeights.current, [index]: size };
+  }
+
+  // Renderer for virtualized rows in `VariableSizeList`.
+  function ListLegendRow({ index, style }: ListChildComponentProps) {
+    // Storing a ref to the row's `ListLegendItem`, so we can get the "real"
+    // height and adjust the height of the row based on it, enabling the dynamic
+    // heights on hover.
+    const rowRef = useRef<HTMLDivElement | null>(null);
+
+    // Adjust row heights when the row being rendered changes.
+    useEffect(() => {
+      if (rowRef.current) {
+        setRowHeight(index, rowRef.current.clientHeight);
+      }
+    }, [rowRef, index]);
+
+    // useCallback is important here to avoid constantly running the useEffect
+    // that calls this in `ListLegendItem`.
+    const handleRowLayoutChange = useCallback(() => {
+      // Handle size changes from hovering on a list item.
+      if (rowRef.current) {
+        setRowHeight(index, rowRef.current.clientHeight);
+      }
+    }, [index]);
+
+    const item = items[index];
+
+    if (!item) {
+      // This shouldn't happen if configured correctly, but covering
+      // the case to appease the type checking and to cover any edge
+      // cases.
+      return null;
+    }
+
+    const originalTop = parseFloat(`${style.top}`);
+
+    return (
+      <div
+        style={{
+          ...style,
+          // Adjust the top position to simulate top padding on the list.
+          top: `${originalTop + LIST_PADDING}px`,
+        }}
+      >
         <ListLegendItem
+          ref={rowRef}
           key={item.id}
           item={item}
+          truncateLabel={truncateLabels}
+          onLayoutChange={handleRowLayoutChange}
           sx={{
-            width: 190,
-            textOverflow: 'ellipsis',
+            // Having an explicit width is important for the ellipsizing to
+            // work correctly. Subtract padding to simulate padding.
+            width: width - LIST_PADDING,
             wordBreak: 'break-word',
-            overflow: truncateLabels ? 'hidden' : 'visible',
```

Comment:
I realized the truncation was broken while I was in here. With this html layout, it needs to happen farther down in the code, which was a little uglier to do, but works as intended.